### PR TITLE
Adds missing breaking changes to the upgrade guide

### PR DESCRIPTION
**Pull request by jlstrater**
This change adds more breaking changes that were previously in the release notes but were absent from this upgrade guide.

Since the document was getting quite long in some sections, I moved some of the content.

I also added links to some of the notable libraries and their changelogs when there are version upgrades.

### CODE REVIEW

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 1]
```diff
@@ -48,24 +49,41 @@ Other notable changes to be aware of that may break your build include:
...
-There is only one change that was not previously deprecated:
+The following breaking changes are not from deprecations, but the result of changes in behavior:
 
  * Optional dependency declarations in POMs no longer influence the version of a dependency that Gradle picks.
 +
 This change is only applicable if you had the `IMPROVED_POM_SUPPORT` preview feature enabled.
  * The evaluation of the `publishing {}` block is no longer deferred until needed but behaves like any other block.
    Please use `afterEvaluate {}` if you need to defer evaluation.
+ * The link:{groovyDslPath}/org.gradle.api.tasks.javadoc.Javadoc.html[`Javadoc`] and link:{groovyDslPath}/org.gradle.api.tasks.javadoc.Groovydoc.html[`Groovydoc`] tasks now delete the destination dir for the documentation before executing. This has been added to remove stale output files from the last task execution.
+ * The <<#java_library_distribution_plugin, Java Library Distribution Plugin>> is now based on the <<#java_library_plugin, Java Library Plugin>> instead of the <<#java_plugin, Java Plugin>>.
```
Comment (jlstrater) [inline subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: It isn't entirely clear to me what the implications of this change are.  Could we reword this to describe what problems might happen as a result of this change?

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 2]
```diff
@@ -131,19 +155,38 @@ Tasks & properties::
...
 Scala & Play::
  * Play 2.2 is no longer supported — please upgrade the version of Play you are using.
  * The `ScalaDocOptions.styleSheet` property has been removed — the Scaladoc Ant task in Scala 2.11.8 and later no longer supports this property.
+Kotlin DSL::
+ * Artifact configuration accessors are now typed `NamedDomainObjectProvider<Configuration>` instead of simply `Configuration`
+ * `PluginAware.apply<T>(to)` was renamed `PluginAware.applyTo<T>(target)`.
++
+Both changes could cause script compilation errors.
++
+See the https://github.com/gradle/kotlin-dsl/releases/tag/v1.0-RC13#breaking-changes[Gradle Kotlin DSL release notes] for more information and how to fix builds broken by the changes described above.
```
Comment (jlstrater) [inline subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: @gradle/kotlin-dsl is this link correct? I just copied it from the release notes.

Comment (eriwen) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: It resolves but will need to be changed before 5.0 GA to whatever Kotlin DSL 1.0.x release

Comment (eriwen) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Scratch that... it doesn't resolve. This does: https://github.com/gradle/kotlin-dsl/releases/tag/v1.0-RC12#breaking-changes

How does our release notes test miss that... I thought it checked all links...

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 1 above)
Comment (eriwen) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: @gradle/dependency-management ☝️

**Comment by eriwen** [comment]
@jlstrater one request with this: could we now include reverting 98ae9f6fb4ede4274a34f41a81b8b8b8edbf3419 in this PR?

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 2 above)
Comment (eskatos) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: 5.0-RC1 will go out with 1.0-RC13, the release notes aren't published yet, they will be Monday, that link is correct. You should see the draft here https://github.com/gradle/kotlin-dsl/releases

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 3]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+ * The JUnit Platform libraries used by the `Test` task have been upgraded from 1.0.3 to 1.3.1.
+ * The Maven Wagon libraries used to access Maven repositories have been upgraded from 2.4 to 3.0.0.
+ * SLF4J has been upgraded from 1.7.16 to https://www.slf4j.org/news.html[1.7.25].
+
+[[rel5.0:dependency_resolution]]
+=== [5.0] Fixes to dependency resolution
+
+Dependency resolutions fixes have been included in this release.
+By definition this could impact the set of resolved dependencies of your build.
+However the fixed issues are mostly about corner cases and combination with recent features and thus should have a limited impact.
```
Comment (pledbrook) [inline subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Sounds wrong. The "...and combination" doesn't fit. Perhaps it's supposed to be "... in combination with...". Also, I would drop the "thus".

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 4]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+
+When a dependency constraint matched a real dependency, it was made part of the graph.
+However if for some reason the dependency was later evicted from the graph, the constraint remained present. Now, when the last non-constraint edge to a dependency disappears, all constraints for that dependency will be properly removed from the graph.
```
Comment (pledbrook) [inline subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: What's this paragraph about? Is it introducing a specific fix? Or is this the only fix that the release made? We need to make a connection with the previous paragraph in some way.

It would also be useful to get a translation of this paragraph for "normal" users :smile: It's a bit hard to grok what it's saying.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 5]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+
+[[rel5.0:changes_to_default_task]]
+=== [5.0] Changes to property factory methods on DefaultTask
+
+Property factory methods on DefaultTask are final
+The property factory methods such as newInputFile() are intended to be called from the constructor of a type that extends DefaultTask. These methods are now final to avoid subclasses overriding these methods and using state that is not initialized.
```
Comment (pledbrook) [inline subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Monospace for methods, properties, fields and type please.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 6]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+By definition this could impact the set of resolved dependencies of your build.
+However the fixed issues are mostly about corner cases and combination with recent features and thus should have a limited impact.
+
+When a dependency constraint matched a real dependency, it was made part of the graph.
+However if for some reason the dependency was later evicted from the graph, the constraint remained present. Now, when the last non-constraint edge to a dependency disappears, all constraints for that dependency will be properly removed from the graph.
+
+[[rel5.0:changes_to_default_task]]
+=== [5.0] Changes to property factory methods on DefaultTask
+
+Property factory methods on DefaultTask are final
```
[the remark this answers is not in the document]
Comment (pledbrook) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: I looks like it may have been a heading in its previous incarnation. As an introductory sentence, it shouldn't definitely say "...are _now_ final."

Also, monospace for type please.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 7]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+The property factory methods such as newInputFile() are intended to be called from the constructor of a type that extends DefaultTask. These methods are now final to avoid subclasses overriding these methods and using state that is not initialized.
+
+Inputs and outputs are not automatically registered
+The Property instances that are returned by these methods are no longer automatically registered as inputs or outputs of the task. The Property instances need to be declared as inputs or outputs in the usual ways, such as attaching annotations such as @OutputFile or using the runtime API to register the property.
```
Comment (pledbrook) [inline subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Oh, and monospace for annotations as well!

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 8]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+    @OutputFile // property needs an annotation
+    final RegularFileProperty outputFile = project.objects.fileProperty()
+}
+
+task myOtherTask {
+    def outputFile = project.objects.fileProperty()
+    outputs.file(outputFile) // or to be registered using the runtime API
+    doLast { ... }
+}
+----
```
Comment (pledbrook) [inline subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: It's not clear what these samples are demonstrating. I would be a bit more explicit with the intro sentence. Something like:

> For example, you could previously use the following syntax and have both `outputFile` instances registered as declared outputs:

I would also link to the user manual section on incremental build somewhere in the preceding paragraph.

Comment (pledbrook) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: The other intro sentence could be something like:

> Now you have to explicitly register `outputFile`, like this:

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 9]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+
+[[rel5.0:jaxb_and_java9]]
+=== [5.0] Gradle now bundles JAXB for Java 9 and above
+
+In order to use S3 backed artifact repositories, it was previously required to add `--add-modules java.xml.bind` to `org.gradle.jvmargs` when running on Java 9 and above.
```
Comment (pledbrook) [inline subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: "S3-backed" (hyphen). Also, replace "it was previously require to add" with "you previously had to add...".

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 10]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+
+Please remove the `--add-modules java.xml.bind` option from `org.gradle.jvmargs`, if set.
+
+[[rel5.0:java_library_distribution_plugin]]
+=== Java Library Distribution Plugin utilizes Java Library Plugin
+
+The <<#java_library_distribution_plugin, Java Library Distribution Plugin>> is now based on the
+<<#java_library_plugin, Java Library Plugin>> instead of the <<#java_plugin, Java Plugin>>.
+
+Additionally the created distribution will contain all artifacts of the `runtimeClasspath` configuration instead of the deprecated `runtime` configuration.
```
Comment (pledbrook) [inline subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: "Additionally, the default distribution created by the plugin will contain...". The "created distribution" is a little unclear, although not particularly ambiguous once you think about it. But we shouldn't require our readers to think hard to understand what is meant.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 11]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+so the `register()` for task `bar` only executes when `foo` is configured.
+
+[source, groovy]
+----
+tasks.register("foo") {
+    tasks.register("bar") // WRONG
+}
+----
+
+To avoid this, Gradle now detects this and prevents modification to the underlying container (through `create` or `register`) when using the new API.
```
[the remark this answers is not in the document]
Comment (pledbrook) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Parentheses for `create()` and `register()`.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 1 above)
Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: This was added through #5976 so maybe @marcphilipp can comment. But I agree that it looks more like an _improvement_ than a _breaking change_.

Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: And I should have read the comments on that PR before answering :wink:. Breaking change it is, though I am not sure how to express it better.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 12]
```diff
@@ -131,19 +155,38 @@ Tasks & properties::
  * You can no longer register invalid inputs and outputs via link:{javadocPath}/org/gradle/api/tasks/TaskInputs.html[TaskInputs] and link:{javadocPath}/org/gradle/api/tasks/TaskOutputs.html[TaskOutputs].
  * The `TaskDestroyables.file()` and `TaskDestroyables.files()` methods have been removed — use link:{javadocPath}/org/gradle/api/tasks/TaskDestroyables.html#register-java.lang.Object++...++-[TaskDestroyables.register()] instead.
  * `SimpleWorkResult` has been removed — use link:{javadocPath}/org/gradle/api/tasks/WorkResult.html#getDidWork--[WorkResult.didWork].
+ * Overriding default tasks <<#deprecations_4.8,deprecated in 4.8>> now produces an error.
++
+Attempting to replace a built-in task will produce an error similar to the following:
+  > Cannot add task 'wrapper' as a task with that name already exists.
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Ack. My mistake.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 13]
```diff
@@ -131,19 +155,38 @@ Tasks & properties::
...
 Scala & Play::
  * Play 2.2 is no longer supported — please upgrade the version of Play you are using.
  * The `ScalaDocOptions.styleSheet` property has been removed — the Scaladoc Ant task in Scala 2.11.8 and later no longer supports this property.
+Kotlin DSL::
+ * Artifact configuration accessors are now typed `NamedDomainObjectProvider<Configuration>` instead of simply `Configuration`
+ * `PluginAware.apply<T>(to)` was renamed `PluginAware.applyTo<T>(target)`.
++
+Both changes could cause script compilation errors.
++
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: That's a good point. I need to look at it a little further. When I tried that the first time, it went all the way flush with the main list and not the sublist.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 14]
```diff
@@ -131,19 +155,38 @@ Tasks & properties::
...
  * The `@DeferredConfigurable` annotation has been removed.
+ * The method `isDeferredConfigurable` was removed from `ExtensionSchema`.
  * `IdeaPlugin.performPostEvaluationActions()` and `EclipsePlugin.performPostEvaluationActions()` have been removed.
  * `The `BroadcastingCollectionEventRegister.getAddAction()` method has been removed with no replacement.
  * The internal `org.gradle.util` package is no longer imported by default.
 +
 Ideally you shouldn't use classes from this package, but, as a quick fix, you can add explicit imports to your build scripts for those classes.
+ * The Tooling API can no longer connect to builds using a Gradle version below Gradle 2.6. The same applies to builds run through TestKit.
+ * Gradle 5.0 requires a minimum Tooling API client version of 3.0. Older client libraries can no longer run builds with Gradle 5.0.
+* The IdeaModule Tooling API model element contains methods to retrieve resources and test resources so those elements were removed from the result of IdeaModule#getSourceDirs() and IdeaModule#getTestSourceDirs().
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Good catch of the copy pasta. Thanks.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 15]
```diff
@@ -131,19 +155,38 @@ Tasks & properties::
...
+ * In previous Gradle versions the `source` filed in `SourceTask` was accessible from subclasses.
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: probably. which team wrote this? It also needs a clause to describe what happens now. I assume it's no longer accessible?

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 16]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+
+All workers, including compilers and test executors, now start with 512m of heap. The previous default was 1/4th of physical memory.
+Large projects may have to increase this setting on the relevant tasks, e.g. `JavaCompile` or `Test`.
+
+[[rel5.0:default_tool_versions]]
+=== [5.0] New default versions for code quality plugins
+
+The default tool versions of the following code quality plugins have been updated:
+
+ * The http://checkstyle.sourceforge.net[Checkstyle] plugin now uses http://checkstyle.sourceforge.net/releasenotes.html#Release_8.12[8.12] instead of 6.19 by default.
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: This was something I was debating and ended up with this syntax. I'm linking to the technologies not the Gradle plugins. Is it confusing if I include the word plugin in the link?

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 17]
```diff
@@ -19,7 +19,7 @@ This chapter provides the information you need to migrate your older Gradle 4.x
 
 We recommend the following steps for all users:
 
- . If you are not already on version 4.10.2, follow the directions <<#changes_5.0,below>> to upgrade your project to 4.10.2.
+ . If you are not already on version 4.10.2, read the sections <<#changes_5.0,below>> for help upgrading your project to 4.10.2.
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: It's not required, but recommended per @eriwen

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 18]
```diff
@@ -48,24 +49,41 @@ Other notable changes to be aware of that may break your build include:
  * A change that means you should <<#rel4.8:configure_internal_tasks,configure existing `wrapper` and `init` tasks>> rather than defining your own.
  * The <<#rel4.8:pom_wildcard_exclusions,honoring of implicit wildcards in Maven POM exclusions>>, which may result in dependencies being excluded that weren't before.
  * A <<#rel4.6:annotation_processor_configuration,change to the way you add Java annotation processors to a project>>.
+ * The <<#rel5.0:default_memory_settings,default memory settings>> have been reduced.
+ * The <<#rel5.0:default_tool_versions,default versions>> of several code quality plugins have been updated.
+ * <<#rel5.0:library_upgrades,Several libraries>> used by Gradle have been updated.
+ * A change in behavior related to <<#rel5.0:dependency_resolution,dependency resolution>> may impact a small number of users.
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Thanks. I'll move it.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 19]
```diff
@@ -48,24 +49,41 @@ Other notable changes to be aware of that may break your build include:
...
  * The evaluation of the `publishing {}` block is no longer deferred until needed but behaves like any other block.
    Please use `afterEvaluate {}` if you need to defer evaluation.
+ * The link:{groovyDslPath}/org.gradle.api.tasks.javadoc.Javadoc.html[`Javadoc`] and link:{groovyDslPath}/org.gradle.api.tasks.javadoc.Groovydoc.html[`Groovydoc`] tasks now delete the destination dir for the documentation before executing. This has been added to remove stale output files from the last task execution.
+ * The <<#java_library_distribution_plugin, Java Library Distribution Plugin>> is now based on the <<#java_library_plugin, Java Library Plugin>> instead of the <<#java_plugin, Java Plugin>>.
++
+Additionally the created distribution will contain all artifacts of the `runtimeClasspath` configuration instead of the deprecated `runtime` configuration.
+
+ * For easier configurability from statically compiled languages such as Java or Kotlin, `CheckstyleReports` and `FindbugsReports` `html` property now return `CustomizableHtmlReport`.
+ * The <<#rel5.0:configuration_avoidance, Configuration Avoidance API>> has been updated to prevent creating and configuring tasks that are never used.
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: I like it. 👍

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 20]
```diff
@@ -193,6 +236,12 @@ Use http://docs.groovy-lang.org/latest/html/documentation/#_spread_operator[Groo
 This will lead to some types annotated according to JSR-305 being treated as nullable where they were treated as non-nullable before. This may lead to compilation errors in the build script. See https://github.com/gradle/kotlin-dsl/releases/tag/v0.17.4[the relevant Kotlin DSL release notes] for details.
  * Error messages will be directed to standard error rather than standard output now, unless a console is attached to both standard output and standard error. This may affect tools that scrape a build's plain console output. Ignore this change if you're upgrading from an earlier version of Gradle.
 
+=== [[deprecations_4.8]] Deprecations
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: I tried that first, but for some reason it didn't work at this level only when it's `==`.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 21]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
 
 == Changes in detail
 
+[[rel5.0:default_memory_settings]]
+=== [5.0] Default memory settings changed
+
+The command line client now starts with 64m of heap instead of 1g.
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: I was wondering that too. I always see it in caps so I'll change it.

[the remark this answers is not in the document]
Comment (eskatos) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Note that we should use `MB` not `Mb`, we are counting bytes (`B`), not bits (`b`).

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 20 above)
[the remark this answers is not in the document]
Comment (pledbrook) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Hmmm...I'm pretty sure I've got it to work even on paragraphs, not just headings. Maybe I didn't. Seems weird. The joys of technology.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 16 above)
[the remark this answers is not in the document]
Comment (pledbrook) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: It's a tricky one. If I wanted to focus on the tool itself, I would use "The plugin for Checkstyle ...", retaining the link on "Checkstyle". Having just the link on "Checkstyle" in the name "Checkstyle Plugin" would look weird.

Keeping "plugin" lowercase runs into the issues that dissuade me from using it, even though the uppercase "Plugin" looks high ceremony and a bit over the top. "plugin" could refer to something that plugs into Checkstyle rather than Gradle. And we don't talk about "the plugin called Checkstyle" because its name is Checkstyle Plugin to distinguish it from the tool it integrates with. I'm not a huge fan of the naming convention, but I couldn't really think of anything better that works in all cases without ambiguity.

[the remark this answers is not in the document]
Comment (pledbrook) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Hope that helps :smile:

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 15 above)
[the remark this answers is not in the document]
Comment (pledbrook) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: See [this commit](https://github.com/gradle/gradle/commit/9e0f9be30d682882119f49ff5926d1e32ec26a27#diff-5267e6b16f7b156ab75dabc8537010ad). It should indeed be "field". It still has getter and setter methods for it, as well as a `source()` method. It's just the field's visibility that has changed.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 1 above)
Comment (marcphilipp) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: Maybe add sth. like the following?

> While it applies the Java Plugin, it behaves slightly different (e.g. it adds the `api` configuration). Thus, make sure to check whether your build behaves as expected after upgrading.

Still kind of vague, though. 😕

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 22]
```diff
@@ -93,6 +112,8 @@ Running Gradle & build environment::
 Working with files::
  * You can no longer cast `FileCollection` objects to other types using the `as` keyword or the `asType()` method.
  * You can no longer pass `null` as the configuration action of link:{javadocPath}/org/gradle/api/file/CopySpec.html#from-java.lang.Object-org.gradle.api.Action-[CopySpec.from(Object, Action)].
+ * For better compatibility with the Kotlin DSL, the property setter no longer accepts `null` as a way
+to reset the property back to its default value. Use `DuplicatesStrategy.INHERIT` instead.
```
[the remark this answers is not in the document]
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: 🤦‍♀️ I removed the first part of that statement.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 1 above)
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: applied @marcphilipp 's change after this line.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 23]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+[[rel5.0:java_library_distribution_plugin]]
+=== Java Library Distribution Plugin utilizes Java Library Plugin
+
+The <<#java_library_distribution_plugin, Java Library Distribution Plugin>> is now based on the
+<<#java_library_plugin, Java Library Plugin>> instead of the <<#java_plugin, Java Plugin>>.
+
+Additionally the created distribution will contain all artifacts of the `runtimeClasspath` configuration instead of the deprecated `runtime` configuration.
+
+[[rel5.0:configuration_avoidance]]
+=== Configuration Avoidance API disallows common configuration errors
```
[the remark this answers is not in the document]
Comment (lacasseio) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: @eriwen It's a change in behavior compared to what users may be used to with the eager APIs. In that sense, it's a breaking change.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc [hunk 24]
```diff
@@ -328,6 +377,149 @@ To override this behavior, you can explicitly declare the configuration to use i
...
+To avoid this, Gradle now detects this and prevents modification to the underlying container (through `create` or `register`) when using the new API.
+
+[[rel5.0:worker_api]]
+=== [5.0] Worker API: working directory of a worker can no longer be set
+
+Since JDK 11 no longer supports changing the working directory of a running process, setting the working directory of a worker via its fork options is now prohibited.
+
+All workers now use the same working directory to enable reuse.
+
+Please pass files and directories as arguments instead.
```
[the remark this answers is not in the document]
Comment (lacasseio) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: I think @ghale did this change. I will let him respond to this.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 4 above)
Comment (jlstrater) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: @gradle/dependency-management ? Can you explain it again for us?

Comment (ljacomet) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: So that was from me 😉 

I think both paragraph should be merged / updated. What about:
> Various issues have been fixed in Gradle's dependency resolution for this release. By definition this could impact the set of resolved dependencies of your build.
> If you have been using the `IMPROVED_POM_SUPPORT` feature preview, playing with constraints or `prefer`, `reject` and other specific version indications, then make sure to take a good look at your dependency resolution results.

The original explanation is indeed hard to parse as the issue stems from multiple steps, having to happen in a specific order for the change in behavior to be observed. I think what I suggest above is easier and users should easily be able to say in which category they belong.

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 23 above)
[the remark this answers is not in the document]
Comment (eriwen) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: @lacasseio Hmm, still feels a bit off. I think we ought to document this in the lazy configuration bit of the user manual instead. Do you agree?

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 11 above)
[the remark this answers is not in the document]
Comment (eriwen) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: @lacasseio can you help explain here?

**Comment by jlstrater** [comment]
There are a couple of open questions for the other teams. Should I leave this open or can we merge it and create issues for the open items?

**Comment by eriwen** [comment]
I vote merge and follow up with other teams for not PRs

File: subprojects/docs/src/docs/userguide/upgrading_version_4.adoc (hunk 23 above)
[the remark this answers is not in the document]
Comment (lacasseio) [reply subprojects/docs/src/docs/userguide/upgrading_version_4.adoc]: You are right. At the moment, we don't have time to document any further and given it's an incubating feature breakage are to be expected. However, those breakage aren't really breakage as they just notify the user about those configuration errors.
